# [LAB-02] NumPy

## 4. 기술 통계량의 이해 (1)

### #01 준비작업

1. 라이브러리 참조

In [2]:
import numpy as np

2. 샘플 데이터 생성

In [5]:
myarr = np.array([98, 72, 80, 64, 88, 65, 79, 59, 81, 74])
myarr

array([98, 72, 80, 64, 88, 65, 79, 59, 81, 74])

### #02 데이터의 위치, 분포를 나타내는 기술 통계량

1. 총 도수

In [4]:
print("총 도수:", len(myarr))

총 도수: 10


2. 최대, 최소, 합계

In [6]:
print("최대값:", np.max(myarr))
print("최소값:", np.min(myarr))
print("합계:", np.sum(myarr))

최대값: 98
최소값: 59
합계: 760


3. 평균

In [7]:
print("평균(average):", np.average(myarr))
print("평균(mean):", np.mean(myarr))

평균(average): 76.0
평균(mean): 76.0


4. 중앙값

4-1. 직접 구하기

In [8]:
size = len(myarr)
sorted_arr = np.sort(myarr)
print("정렬:", sorted_arr)

# 짝수인 경우 가운데 두 수의 평균
if size % 2 == 0:
    p = (size // 2) - 1
    q = p + 1
    median = (sorted_arr[p] + sorted_arr[q]) / 2
else:
    p = size // 2
    median = sorted_arr[p]

print("중앙값", median)

정렬: [59 64 65 72 74 79 80 81 88 98]
중앙값 76.5


4-2. NumPy 사용

In [9]:
print("중앙값:", np.median(myarr))

중앙값: 76.5


5. 1사분위 수

In [11]:
x = np.sort(myarr)              # 데이터 정렬
n = len(x)                      # 총 도수
q1_pos = (n - 1) * 0.25         # 위치 계산(p)
i = int(np.floor(q1_pos))       # 아래쪽 인덱스(i)
f = q1_pos - i                  # 보간비율(f)

# 보간 (interpolation)
Q1 = x[i] + f * (x[i + 1] - x[i])
print("직접 계산한 1사분위수 (Q1):", Q1)

# numpy 내장 함수과 비교
print("Numpy로 계산한 1사분위수:", np.quantile(myarr, 0.25))

직접 계산한 1사분위수 (Q1): 66.75
Numpy로 계산한 1사분위수: 66.75


6. 3사분위 수

In [12]:
x = np.sort(myarr)              # 데이터 정렬
n = len(x)                      # 총 도수
q3_pos = (n - 1) * 0.75         # 위치 계산(p)
i = int(np.floor(q3_pos))       # 아래쪽 인덱스(i)
f = q3_pos - i                  # 보간비율(f)

# 보간 (interpolation)
Q3 = x[i] + f * (x[i + 1] - x[i])
print("직접 계산한 3사분위수 (Q3):", Q3)

# numpy 내장 함수과 비교
print("Numpy로 계산한 3사분위수:", np.quantile(myarr, 0.75))

직접 계산한 3사분위수 (Q3): 80.75
Numpy로 계산한 3사분위수: 80.75


### #04 데이터의 범위, 산포를 나타내는 기초 통계량

1. 데이터 구간

In [13]:
print("데이터 구간:", np.max(myarr - np.min(myarr)))

데이터 구간: 39


2. 사분의 수 구간(IQR)

In [14]:
q3 = np.quantile(myarr, 0.75)
q1 = np.quantile(myarr, 0.25)
print("사분위수 구간(IQR):", q3 - q1)

사분위수 구간(IQR): 14.0


3. 분산, 표준편차

In [15]:
print("분산:", np.var(myarr))
print("표준편차:", np.std(myarr))

분산: 125.2
표준편차: 11.189280584559492


### #05 2차원 배열의 기술 통계량

1. 각 열끼리 계산

In [17]:
mydata = np.array([
    [98, 72, 80, 64, 88],
    [65, 79, 59, 81, 74]
])

s1 = np.sum(mydata, axis=0)
print(type(s1))
s1

<class 'numpy.ndarray'>


array([163, 151, 139, 145, 162])

2. 각 행끼리 계산

In [18]:
mydata = np.array([
    [98, 72, 80, 64, 88],
    [65, 79, 59, 81, 74]
])

s2 = np.sum(mydata, axis=1)
print(type(s2))
s2

<class 'numpy.ndarray'>


array([402, 358])

## 연습문제

### 1. 게임 점수 데이터 분석

당신은 게임 회사의 데이터 분석가입니다.   
두 팀, 'Alpha'와 'Bravo'의 게임 점수 데이터가 주어졌습니다.   
각 팀의 점수에 대한 기술 통계량을 아래 표와 같이 산출하세요.

| 구분            | 점수                                        |
| ------------- | ----------------------------------------- |
| ‘Alpha’ 팀 점수: | [88, 92, 76, 100, 95, 81, 79, 98, 85, 87] |
| ‘Bravo’ 팀 점수: | [65, 99, 78, 72, 80, 91, 75, 82, 89, 70]  |


In [21]:
Alpha = np.array([88, 92, 76, 100, 95, 81, 79, 98, 85, 87])
Bravo = np.array([65, 99, 78, 72, 80, 91, 75, 82, 89, 70])

print("==== Alpha팀 기술통계량 ====")
print("도수:", len(Alpha))
print("평균:", np.mean(Alpha))
print("표준편차:", np.std(Alpha))
print("최소값:", np.min(Alpha))
print("1사분위 수:", np.quantile(Alpha, 0.25))
print("중앙값:", np.median(Alpha))
print("3사분위 수:", np.quantile(Alpha, 0.75))
print("최대값:", np.max(Alpha))

print("==== Bravo팀 기술통계량 ====")
print("도수:", len(Bravo))
print("평균:", np.mean(Bravo))
print("표준편차:", np.std(Bravo))
print("최소값:", np.min(Bravo))
print("1사분위 수:", np.quantile(Bravo, 0.25))
print("중앙값:", np.median(Bravo))
print("3사분위 수:", np.quantile(Bravo, 0.75))
print("최대값:", np.max(Bravo))

==== Alpha팀 기술통계량 ====
도수: 10
평균: 88.1
표준편차: 7.7
최소값: 76
1사분위 수: 82.0
중앙값: 87.5
3사분위 수: 94.25
최대값: 100
==== Bravo팀 기술통계량 ====
도수: 10
평균: 80.1
표준편차: 9.924212815130478
최소값: 65
1사분위 수: 72.75
중앙값: 79.0
3사분위 수: 87.25
최대값: 99


### 2. 교통사고 발생량으로부터 인사이트 발견하기

다음의 데이터는 2024년도 월별 교통사고 건수이다.   
이 데이터에 대해 기술 통계량을 구하고 이를 통해 알 수 있는 사실을 하나 이상 서술하시오.

| 발생월   |     1 |     2 |     3 |     4 |     5 |     6 |     7 |     8 |     9 |    10 |    11 |    12 |
| ----- | ----: | ----: | ----: | ----: | ----: | ----: | ----: | ----: | ----: | ----: | ----: | ----: |
| 사고 건수 | 15129 | 13836 | 15738 | 16300 | 17605 | 16690 | 16252 | 17000 | 16956 | 17720 | 17455 | 15668 |

In [ ]:
import numpy as np

accident = np.array([15129, 13836, 15738, 16300, 17605, 16690, 16252, 17000, 16956, 17720, 17455, 15668])

print("도수:", len(accident))
print("최소값:", np.min(accident))
print("최대값:", np.max(accident))
print("평균:", np.mean(accident))
print("중앙값:", np.median(accident))
print("표준편차:", np.std(accident))
print("1사분위 수:", np.quantile(accident, 0.25))
print("3사분위 수:", np.quantile(accident, 0.75))

도수: 12
최소값: 13836
최대값: 17720
평균: 16362.416666666666
중앙값: 16495.0
표준편차: 1089.3134273732037
1사분위 수: 15720.5
3사분위 수: 17113.75


## 답안

### 2024년 월별 교통사고(traffic_acc) — 통계 기반 인사이트

#### 1) 중심 경향성 — 평균과 중앙값의 관계

- 평균(≈16,362)과 중앙값(16,495)은 매우 근접합니다(차이 ≈ -133). 이로 미루어 전체 분포의 중심은 한 군데에 모여 있으며, 평균이 중앙값보다 작다는 점은 왼쪽으로 약간 긴 꼬리가 존재함을 시사합니다. 이는 데이터가 오른쪽으로 쏠림 현상이 있다는 것을 의미합니다.

#### 2) 변동성(표준편차)과 상대적 불확실성

- 표준편차가 약 1,089로 평균 대비 약 6.7% 수준입니다. 즉 월별 사고 수는 평균을 중심으로 대체로 ±1,000건 내외의 변동을 보이고 있습니다.

#### 3) 사분위수 판단

Q3(≈17,113.75) 이상인 달은 상위 25%로, 주어진 데이터에서는 3개(예: 17,720, 17,605, 17,455 / 5, 10, 11월)가 여기에 해당합니다.

이들 달을 우선적으로 원인 분석(기상·통행량·행사·도로조건 등)하고 예방책을 집중 배치하면 효과적입니다.

12월은 상대적으로 술자리가 많은 시즌이므로 많은 운전자들이 운전을 자제하는 것으로 보입니다.

또한 1월은 눈이 많이 오는 계절이기 때문에 운전을 할 수 없는 경우가 많은 것으로 보입니다.

#### 참고자료

- 11월 고속도로 교통사고 사망자 ‘연중 최다’…2차 사고 주의보 [출처:중앙일보]
  - https://www.joongang.co.kr/article/25379158
- 단풍놀이 가다가…10월 교통사고 사망↑ [출처:KBS]
  - https://news.kbs.co.kr/news/pc/view/view.do?ncd=8384587
- 초겨울인 11월 안개 교통사고 주의！
  - https://www.koroad.or.kr/main/board/6/7476/board_view.do?&cp=25&listType=list&bdOpenYn=Y&bdNoticeYn=N